### PR TITLE
refactor the huge "main" function into smaller functions + module

### PR DESCRIPTION
**Pull request by japaric**
what the PR title says. I ran the ELFs from PR #218 and they print the expected output.

This is the most non-trivial change that may have changed semantics
- before we were extracting *two* RAM regions from the probe-rs registry. The "first one" (returned by the iterator) and the "one that contains the initial stack pointer (initial_sp)". we installed the canary on "the first one" but only if it contained the initial_sp
- now, we only extract one RAM region, the one that contains initial_sp, and install the canary there

I think this still does the same, in practice, but not 100% sure.

### CODE REVIEW

File: src/backtrace/mod.rs [hunk 1]
```diff
@@ -62,3 +53,37 @@ pub(crate) fn print(
...
+                log::error!("the program panicked");
+            }
+            Outcome::Ok => {
+                log::info!("device halted without error");
+            }
+        }
+    }
+}
+
+/// convert outomce into an exit code
```
Comment (jonas-schievink) [inline src/backtrace/mod.rs:81]:
```suggestion
/// Converts `Outcome` to an exit code.
```

File: src/canary.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/// |        |
+/// |        |
+/// +--------+
+/// | canary |
+/// +--------+ -> highest_static_var_address
+/// |        |
+/// | static | (variables, fixed size)
+/// |        |
+/// +--------+ -> lowest RAM address
+/// ```
```
Comment (jonas-schievink) [inline src/canary.rs:28]: Awesome :heart:

File: src/canary.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+///
+/// The whole canary is initialized to `CANARY_VALUE` before the target program is started.
+/// The canary size is 10% of the available stack space or 1 KiB, whichever is smallest.
+///
+/// When the programs ends (due to panic or breakpoint) the integrity canary is checked. If it was
```
Comment (jonas-schievink) [inline src/canary.rs:33]:
```suggestion
/// When the programs ends (due to panic or breakpoint) the integrity of the canary is checked. If it was
```

File: src/canary.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+/// "touched" (any of its bytes != `CANARY_VALUE`) then that is considered to be a *potential* stack
+/// overflow
```
Comment (jonas-schievink) [inline src/canary.rs:35]:
```suggestion
/// overflow.
```

File: src/canary.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+///
+/// The canary is not installed if the program memory layout is "inverted" (stack is *below* the
+/// static variables)
```
Comment (jonas-schievink) [inline src/canary.rs:38]:
```suggestion
/// static variables).
```

**Review by jonas-schievink — COMMENTED** [review]
This looks like a very nice improvement, nice!

I haven't thoroughly reviewed every part of this, but if the test-ELFs work then I'm fine with merging this.

**Comment by japaric** [comment]
bors r=jonas-schievink

**Pull request merged**